storage/allocator.go

```diff
@@ -13,75 +13,291 @@
+// according to number of ranges.
+//
+// When choosing a rebalance target, a random store is selected from
+// amongst the set of stores with capacity within rebalancePctFromMean
+// from the mean.
 type allocator struct {
-	storeFinder FindStoreFunc
-	rand        rand.Rand
+	sync.Mutex
+	gossip        *gossip.Gossip
+	randGen       *rand.Rand
+	deterministic bool                  // Set deterministic for unittests
+	capacityKeys  map[string]struct{}   // Tracks gossip keys used for capacity
+	storeLists    map[string]*storeList // Cache from attributes to storeList
 }
 
-// newAllocator creates a new allocator.
-func newAllocator(f FindStoreFunc) *allocator {
-	return &allocator{
-		storeFinder: f,
-		// TODO(bdarnell): use a real random seed.
-		rand: *rand.New(rand.NewSource(0)),
+// newAllocator creates a new allocator for the store specified by
+// storeID.
+func newAllocator(g *gossip.Gossip) *allocator {
+	a := &allocator{
+		gossip:  g,
+		randGen: rand.New(rand.NewSource(rand.Int63())),
+	}
+	// Callback triggers on any capacity gossip updates.
+	if a.gossip != nil {
+		capacityRegex := gossip.MakePrefixPattern(gossip.KeyCapacityPrefix)
+		a.gossip.RegisterCallback(capacityRegex, a.capacityGossipUpdate)
 	}
+	return a
 }
 
-// allocate returns a suitable store based on the supplied
-// attributes list. If none are available / suitable, returns an
-// error. It uses the allocator's StoreFinder to select the set of
-// available stores matching attributes for missing replicas and picks
-// using randomly weighted selection based on available capacities.
-func (a *allocator) allocate(required proto.Attributes, existingReplicas []proto.Replica) (
-	*proto.StoreDescriptor, error) {
+// getUsedNodes returns a set of node IDs which are already being used
+// to store replicas.
+func getUsedNodes(existing []proto.Replica) map[proto.NodeID]struct{} {
 	// Get a set of current nodes -- we never want to allocate on an existing node.
-	usedNodes := make(map[proto.NodeID]struct{})
-	for _, replica := range existingReplicas {
+	usedNodes := map[proto.NodeID]struct{}{}
+	for _, replica := range existing {
 		usedNodes[replica.NodeID] = struct{}{}
 	}
+	return usedNodes
+}
+
+// storeDescFromGossip retrieves a StoreDescriptor from the specified
+// capacity gossip key. Returns an error if the gossip doesn't exist
+// or is not a StoreDescriptor.
+func storeDescFromGossip(key string, g *gossip.Gossip) (*proto.StoreDescriptor, error) {
+	info, err := g.GetInfo(key)
 
-	stores, err := a.storeFinder(required)
 	if err != nil {
 		return nil, err
 	}
+	storeDesc, ok := info.(proto.StoreDescriptor)
+	if !ok {
+		return nil, fmt.Errorf("gossiped info is not a StoreDescriptor: %+v", info)
+	}
+	return &storeDesc, nil
+}
+
+// capacityGossipUpdate is a gossip callback triggered whenever capacity
+// information is gossiped. It just tracks keys used for capacity
+// gossip.
+func (a *allocator) capacityGossipUpdate(key string, contentsChanged bool) {
```

Comment:
this method is used as a callback, so its required signature is defined elsewhere. I'd +1 a change to `_ bool` though.